op-node/l1/receipts.go

```diff
@@ -13,74 +12,82 @@ import (
 	"github.com/ethereum/go-ethereum/core/types"
 )
 
-// fetchReceipts fetches the receipts of the transactions using RPC batching, verifies if the receipts are complete and correct, and then returns results
-func fetchReceipts(ctx context.Context, block eth.BlockID, receiptHash common.Hash, txs types.Transactions, getBatch batchCallContextFn) (types.Receipts, error) {
-	if len(txs) == 0 {
-		if receiptHash != types.EmptyRootHash {
-			return nil, fmt.Errorf("no transactions, but got non-empty receipt trie root: %s", receiptHash)
+func makeReceiptsFn(block eth.BlockID, receiptHash common.Hash) func(txHashes []common.Hash, receipts []*types.Receipt) (types.Receipts, error) {
+	return func(txHashes []common.Hash, receipts []*types.Receipt) (types.Receipts, error) {
+		if len(receipts) != len(txHashes) {
+			return nil, fmt.Errorf("got %d receipts but expected %d", len(receipts), len(txHashes))
 		}
-		return nil, nil
-	}
-
-	receipts := make([]*types.Receipt, len(txs))
-	receiptRequests := make([]rpc.BatchElem, len(txs))
-	for i := 0; i < len(txs); i++ {
-		receipts[i] = new(types.Receipt)
-		receiptRequests[i] = rpc.BatchElem{
-			Method: "eth_getTransactionReceipt",
-			Args:   []interface{}{txs[i].Hash()},
-			Result: &receipts[i], // receipt may become nil, double pointer is intentional
-		}
-	}
-	if err := getBatch(ctx, receiptRequests); err != nil {
-		return nil, fmt.Errorf("failed to fetch batch of receipts: %v", err)
-	}
-
-	// We don't trust the RPC to provide consistent cached receipt info that we use for critical rollup derivation work.
-	// Let's check everything quickly.
-	logIndex := uint(0)
-	for i, r := range receipts {
-		if r == nil { // on reorgs or other cases the receipts may disappear before they can be retrieved.
-			return nil, fmt.Errorf("receipt of tx %d returns nil on retrieval", i)
-		}
-		if r.TransactionIndex != uint(i) {
-			return nil, fmt.Errorf("receipt %d has unexpected tx index %d", i, r.TransactionIndex)
-		}
-		if r.BlockNumber.Uint64() != block.Number {
-			return nil, fmt.Errorf("receipt %d has unexpected block number %d, expected %d", i, r.BlockNumber, block.Number)
-		}
-		if r.BlockHash != block.Hash {
-			return nil, fmt.Errorf("receipt %d has unexpected block hash %s, expected %s", i, r.BlockHash, block.Hash)
-		}
-		for j, log := range r.Logs {
-			if log.Index != logIndex {
-				return nil, fmt.Errorf("log %d (%d of tx %d) has unexpected log index %d", logIndex, j, i, log.Index)
+		if len(txHashes) == 0 {
+			if receiptHash != types.EmptyRootHash {
+				return nil, fmt.Errorf("no transactions, but got non-empty receipt trie root: %s", receiptHash)
 			}
-			if log.TxIndex != uint(i) {
-				return nil, fmt.Errorf("log %d has unexpected tx index %d", log.Index, log.TxIndex)
+		}
+		// We don't trust the RPC to provide consistent cached receipt info that we use for critical rollup derivation work.
```

Comment:
This logic looks correct after looking at it closely, I dont think it would hurt to modularize it into a function so that it could be unit tested against mainnet data